internal/pkg/scaffold/crd.go

```diff
@@ -141,10 +145,70 @@ func (s *CRD) CustomRender() ([]byte, error) {
 		}
 	}
 
-	setCRDVersions(crd)
+	setCRDStorageVersion(crd)
+	if err := checkCRDVersions(crd); err != nil {
+		return nil, err
+	}
 	return k8sutil.GetObjectBytes(crd, yaml.Marshal)
 }
 
+func runCRDGenerator(rule genall.OutputRule, root string) (err error) {
+	gctx := &genall.GenerationContext{
+		Collector: &markers.Collector{
+			Registry: &markers.Registry{},
+		},
+		Checker:    &loader.TypeChecker{},
+		InputRule:  genall.InputFromFileSystem,
+		OutputRule: rule,
+	}
+	absAPIsDir := filepath.Join(root, ApisDir)
+	gvs, err := k8sutil.ParseGroupVersions(absAPIsDir)
+	if err != nil {
+		return errors.Wrapf(err, "error parsing API group versions from directory %+q", absAPIsDir)
+	}
+	apiDirs := []string{}
+	for g, vs := range gvs {
+		for _, v := range vs {
+			apiDirs = append(apiDirs, filepath.Join(absAPIsDir, g, v))
+		}
+	}
+
+	gctx.Roots, err = loader.LoadRoots(apiDirs...)
+	if err != nil {
+		return errors.Wrapf(err, "error loading API roots %+q", apiDirs)
+	}
+
+	g := crdgen.Generator{}
+	if err := g.RegisterMarkers(gctx.Collector.Registry); err != nil {
+		return errors.Wrap(err, "error registering markers for CRD API versions")
+	}
+	if err := g.Generate(gctx); err != nil {
+		return errors.Wrap(err, "error generating a CRD for API versions")
+	}
+	return nil
```

Comment:
Would it simplify this function to use [`Generators.ForRoots()`](https://github.com/kubernetes-sigs/controller-tools/blob/a653df4cae21d140ed307073ccbfb1d7e245d3a6/pkg/genall/genall.go#L130) and then run [this loop from `Runtime.Run()`](https://github.com/kubernetes-sigs/controller-tools/blob/a653df4cae21d140ed307073ccbfb1d7e245d3a6/pkg/genall/genall.go#L164-L170)

As a follow-up, maybe we should extract that loop into a separate `RunUntilError()` function upstream so we don't have to duplicate it?